optuna/integration/lightgbm_tuner/optimize.py

```diff
@@ -0,0 +1,608 @@
+            metric = metric[-1]
+        elif type(metric) is set:
+            metric = list(metric)[-1]
+        else:
+            raise NotImplementedError
+        valid_sets = self.lgbm_kwargs.get('valid_sets')  # type: Optional[VALID_SET_TYPE]
+
+        if self.lgbm_kwargs.get('valid_names') is not None:
+            if type(self.lgbm_kwargs['valid_names']) is str:
+                valid_name = self.lgbm_kwargs['valid_names']
+            elif type(self.lgbm_kwargs['valid_names']) in [list, tuple]:
+                valid_name = self.lgbm_kwargs['valid_names'][-1]
+            else:
+                raise NotImplementedError
+
+        elif type(valid_sets) is lgb.Dataset:
+            valid_name = 'valid_0'
+
+        elif type(valid_sets) in [list, tuple] and len(valid_sets) > 0:
+            valid_set_idx = len(valid_sets) - 1
+            valid_name = 'valid_{}'.format(valid_set_idx)
+
+        else:
+            raise NotImplementedError
+
+        val_score = booster.best_score[valid_name][metric]
+        return val_score
+
+    def higher_is_better(self):
+        # type: () -> bool
+
+        metric_name = self.lgbm_params.get('metric', 'binary_logloss')
+        return metric_name.startswith(('auc', 'ndcg@', 'map@', 'accuracy'))
+
+    def compare_validation_metrics(self, val_score, best_score):
+        # type: (float, float) -> bool
+
+        if self.higher_is_better():
+            return val_score > best_score
+        else:
+            return val_score < best_score
+
+
+class OptunaObjective(BaseTuner):
+    """Objective for hyperparameter-tuning with Optuna."""
+
+    def __init__(
+            self,
+            target_param_names,  # type: List[str]
+            lgbm_params,  # type: Dict[str, Any]
+            train_set,  # type: lgb.Dataset
+            lgbm_kwargs,  # type: Dict[str, Any]
+            best_score,  # type: float
+            pbar=None,  # type: Optional[tqdm.tqdm]
+    ):
+
+        self.target_param_names = target_param_names
+        self.pbar = pbar
+        self.lgbm_params = lgbm_params
+        self.lgbm_kwargs = lgbm_kwargs
+        self.train_set = train_set
+
+        self.report = []  # type: List[Dict[str, Any]]
+        self.trial_count = 0
+        self.best_score = best_score
+        self.best_booster = None
+        self.action = 'tune_' + '_and_'.join(self.target_param_names)
+
+        self._check_target_names_supported()
+
+    def _check_target_names_supported(self):
+        # type: () -> None
+
+        supported_param_names = [
+            'lambda_l1',
+            'lambda_l2',
+            'num_leaves',
+            'feature_fraction',
+            'bagging_fraction',
+            'bagging_freq',
+            'min_child_samples',
+        ]
+        for target_param_name in self.target_param_names:
+            if target_param_name not in supported_param_names:
+                raise NotImplementedError("Parameter `{}` is not supported for tunning")
```

Comment:
Please add the tailing period at the end of a sentence.

Reply:
Fixed: c73e464